{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,'../model')\n",
    "\n",
    "import pickle as pkl\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from datasets import RealDataset\n",
    "from models import CoordinatedModel\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import hypergeom, binom\n",
    "from functools import reduce\n",
    "import scipy.stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>nsnps</th>\n",
       "      <th colspan=\"2\" halign=\"left\">15 snps</th>\n",
       "      <th>Window</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>EF1 ($p$)</th>\n",
       "      <th>EF2 ($p$)</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Downstream signaling</th>\n",
       "      <td>2/3 (0.063)</td>\n",
       "      <td>0/3 (1.0)</td>\n",
       "      <td>0.1mb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth hormone secretion</th>\n",
       "      <td>0/4 (1.0)</td>\n",
       "      <td>1/4 (0.499)</td>\n",
       "      <td>0.1mb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGF-1 secretion</th>\n",
       "      <td>1/4 (0.499)</td>\n",
       "      <td>1/4 (0.499)</td>\n",
       "      <td>0.1mb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGF-1 serum balance</th>\n",
       "      <td>1/7 (0.708)</td>\n",
       "      <td>3/7 (0.074)</td>\n",
       "      <td>0.1mb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>No pathway</th>\n",
       "      <td>0/1 (1.0)</td>\n",
       "      <td>0/1 (1.0)</td>\n",
       "      <td>0.1mb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ras signaling</th>\n",
       "      <td>1/5 (0.581)</td>\n",
       "      <td>1/5 (0.581)</td>\n",
       "      <td>0.1mb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "nsnps                         15 snps              Window\n",
       "                            EF1 ($p$)    EF2 ($p$)       \n",
       "Downstream signaling      2/3 (0.063)    0/3 (1.0)  0.1mb\n",
       "Growth hormone secretion    0/4 (1.0)  1/4 (0.499)  0.1mb\n",
       "IGF-1 secretion           1/4 (0.499)  1/4 (0.499)  0.1mb\n",
       "IGF-1 serum balance       1/7 (0.708)  3/7 (0.074)  0.1mb\n",
       "No pathway                  0/1 (1.0)    0/1 (1.0)  0.1mb\n",
       "Ras signaling             1/5 (0.581)  1/5 (0.581)  0.1mb"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenos = {\n",
    "    'IGF-1': {\n",
    "        'gwas_file':'./snps/igf1.txt',\n",
    "        'ann_file' : './pathways/igf1.hits.closest.core.txt',\n",
    "        'efa_file' : './pkl/igf1_sinarm-igf1-range_top100.pkl',\n",
    "        'snp_file' : './snps/igf1_sinarm-igf1-range_top100.snps',\n",
    "        'used_file': './snps/sinarm-igf1-range_top100.txt',\n",
    "    },\n",
    "    'Urate': {\n",
    "        'gwas_file': './snps/urate.txt',\n",
    "        'ann_file' : './pathways/urate.hits.closest.core.txt',\n",
    "        'efa_file' : './pkl/urate_sinarm-urate-range_top100.pkl',\n",
    "        'snp_file' : './snps/urate_sinarm-urate-range_top100.snps',\n",
    "        'used_file': './snps/sinarm-urate-range_top100.txt',\n",
    "    },\n",
    "    'Testosterone (Male)': {\n",
    "        'gwas_file': './snps/maletest.txt',\n",
    "        'ann_file' : './pathways/testosterone.males.hits.closest.core.txt',\n",
    "        'efa_file' : './pkl/male_testosterone_sinarm-maletest-range_top100.pkl',\n",
    "        'snp_file' : './snps/male_testosterone_sinarm-maletest-range_top100.snps',\n",
    "        'used_file': './snps/sinarm-maletest-range_top100.txt',\n",
    "    },\n",
    "    'Testosterone (Female)': {\n",
    "        'gwas_file': './snps/femaletest.txt',\n",
    "        'ann_file' : './pathways/testosterone.females.hits.closest.core.txt',\n",
    "        'efa_file' : './pkl/female_testosterone_sinarm-femaletest-range_top100.pkl',\n",
    "        'snp_file' : './snps/female_testosterone_sinarm-femaletest-range_top100.snps',\n",
    "        'used_file': './snps/sinarm-femaletest-range_top100.txt',\n",
    "    }\n",
    "}\n",
    "\n",
    "def enrich(pathway, ref, nsnps, top_snps):\n",
    "    res = dict()\n",
    "    for k,v in pathway.items():\n",
    "        res[k] = 1 -scipy.stats.hypergeom.cdf(v-1, nsnps, ref[k], top_snps)\n",
    "    return pd.Series(res, dtype='float64')\n",
    "\n",
    "\n",
    "def get_efa_df(gwas_file, ann_file, efa_file, snp_file, used_file):\n",
    "    df = pd.read_csv(gwas_file, sep='\\t')\n",
    "    df = df[df['CHROM'] != 'X']\n",
    "    df = df[df['CHROM'] != 'Y']\n",
    "    df = df.sort_values(by = 'P', kind='stable')\n",
    "    df['IDX'] = df['CHROM'].astype('str') + ':' + df['POS'].astype('str') + '_' + df['REF'] + '_' + df['A1']\n",
    "\n",
    "    used = pd.read_csv(used_file, header=None, sep='\\s+')\n",
    "    df = used.merge(df, left_on=3, right_on='IDX')\n",
    "\n",
    "    df['CHROM'] = df['CHROM'].astype('int')\n",
    "    df = df.sort_values(by='POS')\n",
    "    df = df.sort_values(by='CHROM', kind='stable').reset_index()\n",
    "    df = df.loc[:,['CHROM', 'POS', 'REF', 'A1', 'P', 'BETA', 'IDX']]\n",
    "\n",
    "    with open(efa_file, 'rb') as f:\n",
    "        efa = pkl.load(f)\n",
    "\n",
    "    # efa.pathways -= efa.beta/2\n",
    "    df['U1_orig'] = efa.pathways[:,0]\n",
    "    df['U2_orig'] = efa.pathways[:,1]\n",
    "    efa.pathways *= efa.beta/np.abs(efa.beta)\n",
    "    df['U1'] = efa.pathways[:,0]\n",
    "    df['U2'] = efa.pathways[:,1]\n",
    "    df['U1_res'] = efa.pathways[:,0] - (efa.pathways[:,0] + efa.pathways[:,1])/2\n",
    "    df['U2_res'] = efa.pathways[:,1] - (efa.pathways[:,0] + efa.pathways[:,1])/2\n",
    "\n",
    "\n",
    "    snps = pd.read_csv(snp_file, header=None)\n",
    "    df['SNP'] = snps\n",
    "    df['LAMBDA'] = efa.weights[0, 1]\n",
    "\n",
    "    ann = pd.read_csv(ann_file, sep='\\t', header=None, na_values='.')\n",
    "    ann = ann.iloc[:,[3, 7, 8, 9]]\n",
    "    ann.columns = ['IDX', 'GENE', 'PATHWAY', 'DIST']\n",
    "    ann = ann.drop_duplicates(subset = ['IDX', 'PATHWAY'])\n",
    "    ann = ann.merge(df, left_on='IDX', right_on='IDX').loc[:,['IDX', 'GENE', 'PATHWAY', 'DIST',]]\n",
    "\n",
    "    # ann = ann.sort_values(by = 'P')\n",
    "    # ann = ann.drop_duplicates(subset= ['GENE', 'PATHWAY'])\n",
    "\n",
    "    df = df.merge(ann, left_on='IDX', right_on='IDX')\n",
    "    nsnps = len(df['IDX'].unique())\n",
    "    return df, ann, nsnps\n",
    "\n",
    "\n",
    "pheno = 'IGF-1'\n",
    "# pheno =  'Urate'\n",
    "# pheno = 'Testosterone (Male)'\n",
    "# pheno = 'Testosterone (Female)'\n",
    "df, ann, nsnps = get_efa_df(**phenos[pheno])\n",
    "\n",
    "\n",
    "gene_dist = float('inf')\n",
    "gene_dist = 1e5\n",
    "\n",
    "table_list = []\n",
    "\n",
    "for top_snps in [15]:\n",
    "    u1_ann = df.sort_values(by = 'U1', ascending=False).head(top_snps).loc[df['DIST'] < gene_dist].PATHWAY.value_counts()\n",
    "    u2_ann = df.sort_values(by = 'U2', ascending=False).head(top_snps).loc[df['DIST'] < gene_dist].PATHWAY.value_counts()\n",
    "    all_ann = ann[ann['DIST'] < gene_dist].PATHWAY.value_counts()\n",
    "\n",
    "    # print('________________________________________________')\n",
    "    enrichment = pd.DataFrame({\n",
    "        'u1_count': u1_ann, \n",
    "        'u1_p': enrich(u1_ann, all_ann, nsnps, top_snps),\n",
    "        'u2_count': u2_ann,\n",
    "        'u2_p': enrich(u2_ann, all_ann, nsnps, top_snps),\n",
    "        'all': all_ann\n",
    "        }\n",
    "    )\n",
    "\n",
    "    u1_count_str = u1_ann.reindex(all_ann.index).fillna(0).astype(int).astype(str)\n",
    "    u1_count_str = u1_count_str.str.cat(all_ann.astype(str), join = 'outer', sep = '/')\n",
    "    u1_p_str = enrich(u1_ann, all_ann, nsnps, top_snps).reindex(all_ann.index).fillna(1).round(3).astype(str)\n",
    "    u1_p_str =u1_p_str.map('({})'.format)\n",
    "\n",
    "    u2_count_str = u2_ann.reindex(all_ann.index).fillna(0).astype(int).astype(str)\n",
    "    u2_count_str = u2_count_str.str.cat(all_ann.astype(str), join = 'outer', sep = '/')\n",
    "    u2_p_str = enrich(u2_ann, all_ann, nsnps, top_snps).reindex(all_ann.index).fillna(1).round(3).astype(str)\n",
    "    u2_p_str = u2_p_str.map('({})'.format)\n",
    "\n",
    "    enrichment = pd.DataFrame({\n",
    "        'EF1 ($p$)': u1_count_str.str.cat(u1_p_str, join='outer', sep = ' '), \n",
    "        'EF2 ($p$)': u2_count_str.str.cat(u2_p_str, join='outer', sep = ' '),\n",
    "        }\n",
    "    ).fillna('0 (1)').reset_index().assign(nsnps = '%d snps' %top_snps)\n",
    "    \n",
    "    table_list.append(enrichment)\n",
    "\n",
    "\n",
    "table = reduce(lambda df1, df2: df1.merge(df2, \"outer\"), table_list)\n",
    "table = table.set_index(['index', 'nsnps']).stack().unstack([1, 2])\n",
    "table.index.name = None\n",
    "table['Window'] = '{}mb'.format(gene_dist/1e6)\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>A1</th>\n",
       "      <th>P</th>\n",
       "      <th>BETA</th>\n",
       "      <th>IDX</th>\n",
       "      <th>U1_orig</th>\n",
       "      <th>U2_orig</th>\n",
       "      <th>U1</th>\n",
       "      <th>U2</th>\n",
       "      <th>U1_res</th>\n",
       "      <th>U2_res</th>\n",
       "      <th>SNP</th>\n",
       "      <th>LAMBDA</th>\n",
       "      <th>GENE</th>\n",
       "      <th>PATHWAY</th>\n",
       "      <th>DIST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>4</td>\n",
       "      <td>9721358</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.022399</td>\n",
       "      <td>4:9721358_G_A</td>\n",
       "      <td>-0.031017</td>\n",
       "      <td>0.091575</td>\n",
       "      <td>-0.031017</td>\n",
       "      <td>0.091575</td>\n",
       "      <td>-0.061296</td>\n",
       "      <td>0.061296</td>\n",
       "      <td>rs13129134_A</td>\n",
       "      <td>-0.285857</td>\n",
       "      <td>SLC2A9</td>\n",
       "      <td>Solute_transport</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>4</td>\n",
       "      <td>10244668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3.740000e-44</td>\n",
       "      <td>-0.148939</td>\n",
       "      <td>4:10244668_C_T</td>\n",
       "      <td>0.031469</td>\n",
       "      <td>-0.065920</td>\n",
       "      <td>-0.031469</td>\n",
       "      <td>0.065920</td>\n",
       "      <td>-0.048695</td>\n",
       "      <td>0.048695</td>\n",
       "      <td>rs570500361_T</td>\n",
       "      <td>-0.285857</td>\n",
       "      <td>SLC2A9</td>\n",
       "      <td>Solute_transport</td>\n",
       "      <td>88109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>4</td>\n",
       "      <td>10160181</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3.650000e-56</td>\n",
       "      <td>-0.190140</td>\n",
       "      <td>4:10160181_A_G</td>\n",
       "      <td>0.031230</td>\n",
       "      <td>-0.065443</td>\n",
       "      <td>-0.031230</td>\n",
       "      <td>0.065443</td>\n",
       "      <td>-0.048336</td>\n",
       "      <td>0.048336</td>\n",
       "      <td>rs763487892_G</td>\n",
       "      <td>-0.285857</td>\n",
       "      <td>SLC2A9</td>\n",
       "      <td>Solute_transport</td>\n",
       "      <td>3622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>4</td>\n",
       "      <td>9860930</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2.830000e-39</td>\n",
       "      <td>-0.151825</td>\n",
       "      <td>4:9860930_G_A</td>\n",
       "      <td>0.018776</td>\n",
       "      <td>-0.038495</td>\n",
       "      <td>-0.018776</td>\n",
       "      <td>0.038495</td>\n",
       "      <td>-0.028635</td>\n",
       "      <td>0.028635</td>\n",
       "      <td>rs557066340_A</td>\n",
       "      <td>-0.285857</td>\n",
       "      <td>SLC2A9</td>\n",
       "      <td>Solute_transport</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>6</td>\n",
       "      <td>25807603</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2.810000e-210</td>\n",
       "      <td>0.017394</td>\n",
       "      <td>6:25807603_A_G</td>\n",
       "      <td>-0.001466</td>\n",
       "      <td>0.055394</td>\n",
       "      <td>-0.001466</td>\n",
       "      <td>0.055394</td>\n",
       "      <td>-0.028430</td>\n",
       "      <td>0.028430</td>\n",
       "      <td>rs2817188_G</td>\n",
       "      <td>-0.285857</td>\n",
       "      <td>SLC17A1</td>\n",
       "      <td>Solute_transport</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>4</td>\n",
       "      <td>9751995</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.250000e-32</td>\n",
       "      <td>-0.133196</td>\n",
       "      <td>4:9751995_C_T</td>\n",
       "      <td>0.016289</td>\n",
       "      <td>-0.036886</td>\n",
       "      <td>-0.016289</td>\n",
       "      <td>0.036886</td>\n",
       "      <td>-0.026587</td>\n",
       "      <td>0.026587</td>\n",
       "      <td>rs528465073_T</td>\n",
       "      <td>-0.285857</td>\n",
       "      <td>SLC2A9</td>\n",
       "      <td>Solute_transport</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM       POS REF A1              P      BETA             IDX   U1_orig  \\\n",
       "32      4   9721358   G  A   0.000000e+00  0.022399   4:9721358_G_A -0.031017   \n",
       "39      4  10244668   C  T   3.740000e-44 -0.148939  4:10244668_C_T  0.031469   \n",
       "38      4  10160181   A  G   3.650000e-56 -0.190140  4:10160181_A_G  0.031230   \n",
       "34      4   9860930   G  A   2.830000e-39 -0.151825   4:9860930_G_A  0.018776   \n",
       "54      6  25807603   A  G  2.810000e-210  0.017394  6:25807603_A_G -0.001466   \n",
       "33      4   9751995   C  T   1.250000e-32 -0.133196   4:9751995_C_T  0.016289   \n",
       "\n",
       "     U2_orig        U1        U2    U1_res    U2_res            SNP    LAMBDA  \\\n",
       "32  0.091575 -0.031017  0.091575 -0.061296  0.061296   rs13129134_A -0.285857   \n",
       "39 -0.065920 -0.031469  0.065920 -0.048695  0.048695  rs570500361_T -0.285857   \n",
       "38 -0.065443 -0.031230  0.065443 -0.048336  0.048336  rs763487892_G -0.285857   \n",
       "34 -0.038495 -0.018776  0.038495 -0.028635  0.028635  rs557066340_A -0.285857   \n",
       "54  0.055394 -0.001466  0.055394 -0.028430  0.028430    rs2817188_G -0.285857   \n",
       "33 -0.036886 -0.016289  0.036886 -0.026587  0.026587  rs528465073_T -0.285857   \n",
       "\n",
       "       GENE           PATHWAY   DIST  \n",
       "32   SLC2A9  Solute_transport      0  \n",
       "39   SLC2A9  Solute_transport  88109  \n",
       "38   SLC2A9  Solute_transport   3622  \n",
       "34   SLC2A9  Solute_transport      0  \n",
       "54  SLC17A1  Solute_transport      0  \n",
       "33   SLC2A9  Solute_transport      0  "
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values(by = 'U1_res', ascending=True).head(top_snps).loc[df['DIST'] < gene_dist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{llllllll}\n",
      "\\toprule\n",
      "nsnps & \\multicolumn{2}{l}{10 snps} & \\multicolumn{2}{l}{15 snps} & \\multicolumn{2}{l}{20 snps} &  Window \\\\\n",
      "{} &     EF1 ($p$) &     EF2 ($p$) &     EF1 ($p$) &      EF2 ($p$) &     EF1 ($p$) & \\multicolumn{2}{l}{EF2 ($p$)} \\\\\n",
      "\\midrule\n",
      "Purine_metabolism &    0/39 (1.0) &  1/39 (0.995) &  3/39 (0.979) &   3/39 (0.979) &   6/39 (0.89) &   5/39 (0.961) &  10.0mb \\\\\n",
      "Solute_transport  &  8/34 (0.003) &    9/34 (0.0) &  9/34 (0.026) &  11/34 (0.001) &  9/34 (0.194) &  13/34 (0.002) &  10.0mb \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-175-072ba65bb0cf>:1: FutureWarning: In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.\n",
      "  print(table.to_latex(escape=False))\n"
     ]
    }
   ],
   "source": [
    "print(table.to_latex(escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('./gwas/igf1_pathways.tsv', sep = '\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 ('py38')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "8f6a110ed188c971830a4fd36635f906538729ac5ffb7c33f242f3e8c395fc3a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
